### PR TITLE
Add ldap support to vault container in docker dev environment

### PR DESCRIPTION
**Pull request by thedoubl3j**
##### SUMMARY

This PR will configure the vault container with LDAP and userpass auth_methods for dev stack. 

##### ISSUE TYPE

 - Bug, Docs Fix or other nominal change

##### COMPONENT NAME

 - Other

##### AWX VERSION

```
devel
```

##### ADDITIONAL INFORMATION

To test this change, pull this branch down and  run:

```
COMPOSE_TAG=devel LDAP=true VAULT=true make docker-compose
```
1. Once the cluster is up, you will need to unseal vault, you can do that manually with the keys found `_sources` in docker compose or you can run the unseal playbook.
```
ansible-playbook tools/docker-compose/ansible/unseal_vault.yml
```
2. Next, plumb the vault container. This playbook will also need you to export your admin creds to your awx instance since we are making changes to awx as well
```
export CONTROLLER_USERNAME=admin
export CONTROLLER_PASSWORD=<password>
ansible-playbook tools/docker-compose/ansible/plumb_vault.yml -e enable_ldap=true
```
3. Once those playbooks are done, you will now have credentials to access both Userpass and LDAP credentials in AWX that were created from a custom vault credential. To test these credentials, you will need to either write a dummy playbook that will print the var that is configured for the test creds (the_secret_from_vault) or you can use a test one @djyasin wrote to test original feature which I manipulated for this demo here: https://github.com/thedoubl3j/Playbooks. 
4. Once you have your playbook, create a project with either yours or the test source material and create a job template with the relative playbook and demo inventory selected. For the job template credential, you will need to select "Vault Custom Cred Type" from the type drop down and then select the method you want to test, userpass or ldap.
5. Save the Job Template and launch it. Given the credential selected, you should see one of two outputs printed to the screen
- LDAP
```
TASK [Print HashiVault Secret] *************************************************
ok: [localhost] => {
    "msg": "this_is_the_ldap_secret_value"
}
```
- UserPass
```
TASK [Print HashiVault Secret] *************************************************
ok: [localhost] => {
    "msg": "this_is_the_userpass_secret_value"
}
```

### CODE REVIEW

File: tools/docker-compose/ansible/roles/sources/templates/ldap.ldif.j2 [hunk 1]
```diff
@@ -84,3 +84,15 @@ objectClass: top
 objectClass: groupOfNames
 member: cn=awx_ldap_org_admin,ou=users,dc=example,dc=org
 
+{% if enable_ldap|bool and enable_vault|bool %}
+dn: cn=awx_ldap_vault,ou=users,dc=example,dc=org
+mail: admin@example.org
+sn: LdapAdmin
+cn: awx_ldap_admin
```
Comment (TheRealHaoLiu) [inline tools/docker-compose/ansible/roles/sources/templates/ldap.ldif.j2]: use a different name here since awx_ldap_admin already exist

Comment (TheRealHaoLiu) [reply tools/docker-compose/ansible/roles/sources/templates/ldap.ldif.j2]: something like hashicorp_ldap_admin?

Comment (thedoubl3j) [reply tools/docker-compose/ansible/roles/sources/templates/ldap.ldif.j2]: Will do. Probably a symptom of copy pasta here as I was walking out the door. Will change it tomorrow

**Review by TheRealHaoLiu — CHANGES_REQUESTED** [review]
```
linters: commands[2]> yamllint -s .
./tools/docker-compose/ansible/roles/vault/defaults/main.yml
  9:11      warning  too few spaces before comment  (comments)
  9:12      warning  missing starting space in comment  (comments)
  10:11     warning  too few spaces before comment  (comments)
  10:12     warning  missing starting space in comment  (comments)

./tools/docker-compose/ansible/roles/vault/tasks/plumb.yml
  97:16     error    trailing spaces  (trailing-spaces)
  98:16     error    trailing spaces  (trailing-spaces)
  107:48    error    trailing spaces  (trailing-spaces)

./tools/docker-compose/ansible/roles/vault/tasks/initialize.yml
  104:17    error    wrong indentation: expected 14 but found 16  (indentation)
  188:17    error    wrong indentation: expected 14 but found 16  (indentation)
  194:1     error    trailing spaces  (trailing-spaces)
  205:19    error    wrong indentation: expected 16 but found 18  (indentation)
  ```

**Comment by TheRealHaoLiu** [comment]
@thedoubl3j in your VSCode u can have it automatically trim trailing whitespace

open command pallet and type `> Trim Trailing White Spaces`

File: tools/docker-compose/ansible/roles/vault/tasks/initialize.yml [hunk 2]
```diff
@@ -92,6 +92,118 @@
...
+          validate_certs: false
+          token: "{{ Initial_Root_Token }}"
+
+        - name: Create userpass engine
+          flowerysong.hvault.engine:
+            path: "userpass_engine"
+            type: "kv"
+            vault_addr: "{{ vault_addr_from_host }}"
+            validate_certs: false
+            token: "{{ Initial_Root_Token }}"
```
Comment (TheRealHaoLiu) [inline tools/docker-compose/ansible/roles/vault/tasks/initialize.yml]: kv engine for userpass user should always exist regardless if LDAP is enabled

File: tools/docker-compose/ansible/roles/vault/tasks/initialize.yml [hunk 3]
```diff
@@ -92,6 +92,118 @@
...
+
+        - name: Create a ldap secret
+          flowerysong.hvault.kv:
+          mount_point: "ldap/my_root"
+          key: "my_folder"
+          value:
+            my_key: "this_is_the_secret_value"
+          vault_addr: "{{ vault_addr_from_host }}"
+          validate_certs: false
+          token: "{{ Initial_Root_Token }}"
```
Comment (TheRealHaoLiu) [inline tools/docker-compose/ansible/roles/vault/tasks/initialize.yml]:
```
TASK [Initialize vault] ********************************************************************************************************************************************************************************************************************************************************
ERROR! conflicting action statements: flowerysong.hvault.kv, mount_point

The error appears to be in '/Volumes/case-sensitive-volume/projects/src/github.com/TheRealHaoLiu/awx/tools/docker-compose/ansible/roles/vault/tasks/initialize.yml': line 119, column 11, but may
be elsewhere in the file depending on the exact syntax problem.

The offending line appears to be:


        - name: Create a ldap secret
          ^ here
```

File: tools/docker-compose/ansible/roles/vault/tasks/initialize.yml (hunk 2 above)
Comment (TheRealHaoLiu) [reply tools/docker-compose/ansible/roles/vault/tasks/initialize.yml]: same for the rest of the userpass related things

File: tools/docker-compose/ansible/roles/vault/tasks/plumb.yml [hunk 4]
```diff
@@ -72,11 +70,77 @@
...
+    inputs:
+      api_version: "v1"
+      default_auth_path: "ldap"
+      kubernetes_role: ""
+      namespace: ""
+      token: "{{ Initial_Root_Token }}"
+      url: "{{ vault_addr_from_container }}"
+      username: 
+      password: 
+  register: vault_ldap_cred
```
Comment (TheRealHaoLiu) [inline tools/docker-compose/ansible/roles/vault/tasks/plumb.yml]: conditional on if LDAP is enabled

File: tools/docker-compose/ansible/roles/vault/tasks/plumb.yml [hunk 5]
```diff
@@ -72,11 +70,77 @@
...
+    injectors:
+      extra_vars:
+        the_secret_from_vault: "{{ '{{' }} password {{ '}}' }}"
+    inputs:
+      fields:
+        - type: "string"
+          id: "password"
+          label: "Password"
+          secret: true
+  register: custom_vault_ldap_cred_type
```
Comment (TheRealHaoLiu) [inline tools/docker-compose/ansible/roles/vault/tasks/plumb.yml]: 1 demo custom credential type for all 3 should be fine since its the same

**Review by djyasin — APPROVED** [review]
I was able to successfully retrieve both the userpass and ldap secrets via these steps!

File: Makefile [hunk 6]
```diff
@@ -538,7 +538,8 @@ docker-compose: awx/projects docker-compose-sources
 	ansible-galaxy install --ignore-certs -r tools/docker-compose/ansible/requirements.yml;
 	ansible-playbook -i tools/docker-compose/inventory tools/docker-compose/ansible/initialize_containers.yml \
 	    -e enable_vault=$(VAULT) \
-	    -e vault_tls=$(VAULT_TLS);
+	    -e vault_tls=$(VAULT_TLS) \
+		-e enable_ldap=$(LDAP);
```
Comment (TheRealHaoLiu) [inline Makefile]:
```suggestion
		-e enable_ldap=$(LDAP);
```
```suggestion
	    -e vault_tls=$(VAULT_TLS) \
	    -e enable_ldap=$(LDAP);
```

File: tools/docker-compose/ansible/roles/sources/templates/ldap.ldif.j2 [hunk 7]
```diff
@@ -41,6 +41,20 @@ objectClass: inetOrgPerson
...
+sn: LdapVaultAdmin
+cn: {{ vault_ldap_username }}
+objectClass: top
+objectClass: person
+objectClass: organizationalPerson
+objectClass: inetOrgPerson
+userPassword: {{ vault_ldap_password }}
+givenName: awx
+{% endif %}
+
```
Comment (TheRealHaoLiu) [inline tools/docker-compose/ansible/roles/sources/templates/ldap.ldif.j2]: move to end of file

File: tools/docker-compose/ansible/roles/vault/tasks/plumb.yml [hunk 8]
```diff
@@ -51,32 +50,127 @@
...
+    controller_host: "{{ awx_host }}"
+    controller_username: admin
+    controller_password: "{{ admin_password }}"
+    validate_certs: false
+    metadata:
+      auth_path: ""
+      secret_backend: "userpass_engine"
+      secret_key: "my_key"
+      secret_path: "userpass_root/userpass_secret"
+      secret_version: ""
```
Comment (TheRealHaoLiu) [inline tools/docker-compose/ansible/roles/vault/tasks/plumb.yml]:
```suggestion
      secret_version: ""

```

Comment (TheRealHaoLiu) [reply tools/docker-compose/ansible/roles/vault/tasks/plumb.yml]: new line at end of file...

we should add some lint-ing to this project...

**Review by TheRealHaoLiu — APPROVED** [review]

**Pull request merged**